lib/fastly_nsq/listener.rb

```diff
@@ -1,117 +1,53 @@
-      @consumer     = consumer || FastlyNsq::Consumer.new(topic: topic, channel: channel)
-      @done         = false
-      @logger       = options.fetch :logger, FastlyNsq.logger
-      @manager      = options[:manager] || FastlyNsq::Manager.new
-      @preprocessor = options[:preprocessor]
-      @processor    = processor
-      @thread       = nil
-      @topic        = topic
-    end
+    raise ArgumentError, "processor #{processor.inspect} does not respond to #call" unless processor.respond_to?(:call)
+    raise ArgumentError, "priority #{priority.inspect} must be a Integer" unless priority.is_a?(Integer)
 
-    def identity
-      {
-        consumer:     @consumer,
-        manager:      @manager,
-        preprocessor: @preprocessor,
-        processor:    @processor,
-        topic:        @topic,
-      }
-    end
+    @channel      = channel
+    @logger       = logger
+    @preprocessor = preprocessor
+    @processor    = processor
+    @topic        = topic
+    @priority     = priority
 
-    def reset_then_dup
-      reset
-      dup
-    end
+    @consumer = consumer || FastlyNsq::Consumer.new(topic: topic,
+                                                    connect_timeout: connect_timeout,
+                                                    channel: channel,
+                                                    queue: FastlyNsq::Feeder.new(self, priority))
 
-    def start
-      @logger.info { "> Listener Started: topic #{@topic}" }
-      @thread ||= safe_thread('listener', &method(:go))
-    end
-
-    def go(run_once: false)
-      until @done
-        next_message do |message|
-          log message
-          preprocess message
-          @processor.process message
-        end
-
-        terminate if run_once
-      end
-
-      @manager.listener_stopped(self)
-    rescue FastlyNsq::Shutdown
-      @manager.listener_stopped(self)
-    rescue Exception => e # rubocop:disable Lint/RescueException
-      @logger.error e.inspect
-      @manager.listener_killed(self)
-    end
-
-    def status
-      @thread.status if @thread
-    end
-
-    def terminate
-      @done = true
-      cleanup
-      return unless @thread
-      @logger.info "< Listener TERM: topic #{@topic}"
-      # Interrupt a Consumer blocking in pop with no messages otherwise it will never shutdown
-      @thread.raise FastlyNsq::Shutdown if @consumer.empty?
-    end
-
-    def kill
-      @done = true
-      cleanup
-      return unless @thread
-      @logger.info "< Listener KILL: topic #{@topic}"
-      @thread.raise FastlyNsq::Shutdown
-    end
-
-    private
+    FastlyNsq.manager.add_listener(self)
+  end
 
-    def log(message)
-      @logger.info "[NSQ] Message received on topic [#{@topic}]: #{message}" if @logger
-    end
+  def call(nsq_message)
+    message = FastlyNsq::Message.new(nsq_message)
+    log message
+    preprocessor&.call(message)
+    result = processor.call(message)
+    nsq_message.finish if result
```

Comment:
I don't think there's a functional difference here, but this is calling `finish` on the nsq_message instance instead of previous being called on the fastlynsq::message instance.